src/entropic/sinkhorn.jl

```diff
@@ -209,3 +222,113 @@ function sinkhorn2(μ, ν, C, ε, alg::Sinkhorn; regularization=false, plan=noth
+measures `μ` and `ν` with respect to a common cost matrix `C`,
+entropic regularization parameter `ε` and algorithm `alg`. 
+
+For `regularization = true`, the Sinkhorn Divergence is that of [^FeydyP19] and is computed as:
+```math
+\\operatorname{S}_{ε}(μ,ν) := \\operatorname{OT}_{ε}(μ,ν)
+- \\frac{1}{2}(\\operatorname{OT}_{ε}(μ,μ) + \\operatorname{OT}_{ε}(ν,ν)),
+```
+where ``\\operatorname{OT}_{ε}(μ,ν)``, ``\\operatorname{OT}_{ε}(μ,μ)`` and
+``\\operatorname{OT}_{ε}(ν,ν)`` are the entropically regularized optimal transport loss 
+between `(μ,ν)`, `(μ,μ)` and `(ν,ν)`, respectively.
+
+For `regularization = false`, the Sinkhorn Divergence is that of [^GPC18] and is computed as:
+```math
+\\operatorname{S}_{ε}(μ,ν) := \\operatorname{W}_{ε}(μ,ν)
+- \\frac{1}{2}(\\operatorname{W}_{ε}(μ,μ) + \\operatorname{W}_{ε}(ν,ν)),
+```
+where ``\\operatorname{W}_{ε}(μ,ν)``, ``\\operatorname{W}_{ε}(μ,μ)`` and
+``\\operatorname{W}_{ε}(ν,ν)`` are the entropically regularized optimal transport *cost*
+between `(μ,ν)`, `(μ,μ)` and `(ν,ν)`, respectively, defined as 
+```math
+\\operatorname{W}_{ε}(μ, ν) = \\langle C, γ^\\star \\rangle
+```
+where ``γ^\\star`` is the entropy-regularised transport plan between `μ` and `ν`. 
+The default algorithm for computing the term ``\\operatorname{OT}_{ε}(μ, ν)`` is the `SinkhornGibbs` algorithm.
+For the terms ``\\operatorname{OT}_{ε}(μ, μ)`` and ``\\operatorname{OT}_{ε}(ν, ν)``, the symmetric fixed point iteration of [^FeydyP19] is used. 
+Alternatively, a pre-computed optimal transport `plan` between `μ` and `ν` may be provided.  
+
+[^GPC18]: Aude Genevay, Gabriel Peyré, Marco Cuturi, Learning Generative Models with Sinkhorn Divergences, Proceedings of the Twenty-First International Conference on Artficial Intelligence and Statistics, (AISTATS) 21, 2018
+[^FeydyP19]: Jean Feydy, Thibault Séjourné, François-Xavier Vialard, Shun-ichi Amari, Alain Trouvé, and Gabriel Peyré. Interpolating between optimal transport and mmd using sinkhorn divergences. In The 22nd International Conference on Artificial Intelligence and Statistics, pages 2681–2690. PMLR, 2019.
+See also: [`sinkhorn2`](@ref)
+"""
+function sinkhorn_divergence(
+    μ::AbstractVecOrMat,
+    ν::AbstractVecOrMat,
+    C,
+    ε,
+    alg::Sinkhorn=SinkhornGibbs(),
+    algμ::Sinkhorn=SymmetricSinkhornGibbs(),
+    algν::Sinkhorn=SymmetricSinkhornGibbs();
+    regularization=true,
+    plan=nothing,
+    kwargs...,
+)
+    OTμν, OTμ, OTν = if (regularization == true) && (plan === nothing)
+        sinkhorn_loss(μ, ν, C, ε, alg; kwargs...),
+        sinkhorn_loss(μ, C, ε, algμ; kwargs...),
+        sinkhorn_loss(ν, C, ε, algν; kwargs...)
+    else
+        sinkhorn2(μ, ν, C, ε, alg; plan=plan, regularization=false, kwargs...),
+        sinkhorn2(μ, C, ε, algμ; plan=nothing, regularization=false, kwargs...),
+        sinkhorn2(ν, C, ε, algν; plan=nothing, regularization=false, kwargs...)
+    end
+    return max.(0, OTμν .- (OTμ .+ OTν) / 2)
```

Comment:
Is `OTμν .- (OTμ .+ OTν) / 2` guaranteed to be non-negative even if the regularization terms are not included (as in Genevay et al.)?

Reply:
I don't think it's guaranteed due to numerics. Otherwise, yes, it's theoretically always greater or equal than zero.

Reply:
Sure, Feydy et al. showed non-negativity for the divergence that includes the regularization terms. But I haven't seen a statement or proof for the non-negativity of the divergence that does not include the regularization terms in Genevay et al.'s paper (they mentioned "positivity is yet to be proved" in the main text). Is it proven nowadays?

Reply:
Yes, it's proven. Actually, it's proven that this is an actual metric.

Reply:
> Yes, it's proven. Actually, it's proven that this is an actual metric.

Let me find the paper and I'll share here.

Reply:
Actually, the proof is in the Feydy's paper. It's Theorem 1.

Reply:
I just saw that it seems that Feydy is incorporating the regularization term. But, I was pretty sure he didn't. I'll double check. Just a min.

Reply:
I do wonder whether this could cause issues with differentiability... however I guess the inputs that would fall afoul of `max` would be vanishingly few.

Reply:
I do believe that it's always positive. But indeed it's not proven. I'll try to figure it out, it does not look very complicated.